### PR TITLE
Automatically delete Cloudinary's user uploaded images in articles

### LINKED ISSUE #13936: Automatically Delete Images from Datastore
# Automatically Delete Images from Datastore
**Note:** This was originally an RFC created by Andy Zhao. I am adding it as an issue in our public repo so that the MLH Fellows can work on it 😄

## Summary
When an article or collection (series) is deleted, their images persist in our datastore. Ideally we delete them automatically whenever the model instance is deleted.

## Motivation
By auto deleting our images, this will help with GDPR and user data ownership/deletion requests. Also, less data stored means less to pay for.

## Estimated size
S

## Design/Implementation
First, some context. Most of our images that are handled by Carrierwave, or mounted, are automatically deleted as a part of Carrierwave's implementation. The two models that do not do this are the Article and Collection (series) models, as their image columns take in any URL. Because they don't, we'll be directly sending an API call via the fog gem to delete images accordingly.

The basic steps are:

1. Add an `after_destroy_commit` callback to the models
2. Gather all the images and pass it along to a worker
3. The worker calls a service object to delete the images
4. Service object deletes the images via API call to AWS
5. 
First, the callback will be added. For example, in `app/models/article.rb`:

``` ruby 
class Article < ApplicationRecord
  IMG_MARKDOWN_REGEX = /!\[.*\]\(#{ApplicationConfig["APP_PROTOCOL"]}#{ApplicationConfig["AWS_BUCKET_NAME"]}.s3.amazonaws.com\/(uploads\/articles.*)\)/.freeze

	after_destroy_commit: :delete_images
  
  # ...
  
  private
  
  def delete_images
    image_paths = article.body_markdown.scan(IMG_MARKDOWN_REGEX).flatten
    main_image_presence = article.main_image&.scan(IMG_MARKDOWN_REGEX)
    image_paths << main_image_presence if main_image_presence # this isn't quite right but you get the idea
    DeleteImageWorker.perform_async(image_paths)
  end
end
```
We did the second step of gathering all the images in the above example.

Third, the worker will call the delete image service:

``` ruby 
# app/workers/images/delete_worker.rb

module Images
  class DeleteWorker
    include Sidekiq::Worker
    sidekiq_options queue: :low_priority, retry: 5

    def perform(image_paths)
      Images::Delete.call(image_paths)
    end
  end
end
```
Third, the service object deletes all the images:
``` ruby
# app/services/images/delete.rb

module Images
  class Delete
    attr_accessor :image_paths

    def initialize(image_paths)
      @image_paths = image_paths
    end

    def self.call(...)
      new(...).call
    end

    def call
      storage = Fog::Aws::Storage.new(
        region: ApplicationConfig["AWS_UPLOAD_REGION"],
        aws_access_key_id: ApplicationConfig["AWS_ID"],
        aws_secret_access_key: ApplicationConfig["AWS_SECRET"]
      )

      storage.delete_multiple_objects(ApplicationConfig["AWS_BUCKET_NAME"], image_paths)
      # returns an Excon::Response object, for ex:
      # <Excon::Response
      #  @data={
      #  		body: { "DeleteResult" => [
      #				{"Deleted" => {"Key" => "uploads/articles/filepath1.png"} },
      #				{"Deleted" => {"Key" => "uploads/articles/filepath2.png"} },
      #     ]},
      #			@status: 200>
    rescue => Excon::Response::BadRequest # 400 error
      # log to DataDog
    end
  end
end
```
## Definition of Done
1. An article's `main_image` and images in the body are deleted appropriately.
2. A collection's `main_image` is deleted when it's destroyed.

## Drawbacks
Like any additional processing, this will add additional code complexity to multiple delete paths.

Another drawback in performance is the fact that AWS S3 does not support getting files in a batch and allowing you to delete it in one API call. Realistically, we don't have articles that have 10+ images, but for those that do the worker may take a while to complete.

This issue does not cover deleting cached files.

## Unsolved Questions
- [SOLVED] How do we verify this implementation in a development environment? Will the developer have to get an AWS account? A: I believe you will need to have an AWS account to test this, but I have personally verified this method in production console . 
- What about cached images through Cloudinary? And Imageproxy/Openresty/Nginx?
   - This issue does not try to delete cached images. For Cloudinary, we are limited by 5000 API calls per month, and we'll need to explore a different way of handling this.
- [SOLVED] How do we test deleting cached images in a development environment? A: Testing for openresty can be done via https://github.com/forem/lfss

## Relevant Skills
- Rails backend

### PR DESCRIPTION
**Pull request by squarebat**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

Currently, for articles and collections, the user uploaded images persist in the database, even after the article/collection using that image is deleted. This PR adds the functions required to delete user uploaded images (to cloudinary) in the article if any. 
   
## Related Tickets & Documents
Resolves #13936 

## Added/updated tests?

- [x] Yes
- [ ] No
- [ ] I need help with writing tests : [solved] I am unsure on how to write tests concerning images being uploaded to AWS. I couldn't figure out how to build an article object that would upload an image to datastore and then copy the image url to article's markdown.

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://admin.forem.com/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

### CODE REVIEW

**Comment by squarebat** [comment]
I have added the test to verify automatic deletion. It successfully deletes images in the public folder. Not that I have commented out the `IMG_MARKDOWN_REGEX` in `article.rb` which needs to be used to run the tests locally.

I checked out the series (collection) feature on dev.to and it does not seem to have any images that are not part of the individual articles. I checked out series by other users too, none have a series main_image. I may be wrong though, since the collections schema does have a main_image column. Either way, there is no way to delete entire collections, so I don't think any modifications are required there. Is there a UX way to add a main_image to series?

**Comment by squarebat** [comment]
I am unsure why the test is failing. It will pass for the localhost URL, but it should pass for the AWS URL as well. Or are test images not uploaded to AWS?

**Comment by maestromac** [comment]
@squarebat looks like unrelated flaky spec. Try merging main and push up another commit

File: app/models/article.rb
Comment (rhymes) [inline app/models/article.rb:31]: I don't think we should keep this in the code, what is the purpose of this regexp if it stays commented in the code?

File: app/models/article.rb
Comment (rhymes) [inline app/models/article.rb:32]: This line is super long, it's odd that rubocop didn't flag it during "git commit": 

Can you run `bundle exec rubocop -a app/models/article.rb` and fix the remaining issue(s)?

The `Layout/LineLength` can be quieted with: `IMG_MARKDOWN_REGEX = %r{...} # rubocop:disable Layout/LineLength` (the comment has to be inline).

That said, I don't think we should hardcode the path here, you can get it from `ArticleImageUploader.new.store_dir`

File: spec/uploaders/article_image_uploader_spec.rb
Comment (rhymes) [inline spec/uploaders/article_image_uploader_spec.rb:104]: Same thing about leaving this commented code here

**Review by rhymes — COMMENTED** [review]
Thanks for tackling this @squarebat!

Am I wrong to assume that this will only work on DEV?

Forem Cloud installations and Forem Self Host use imgproxy and the markdown of the articles doesn't contain the `s3.amazonaws.com` path, so those images won't be detected. 

Will that be part of another PR?

File: app/models/article.rb
Comment (squarebat) [reply app/models/article.rb:31]: I left that in to verify image deletion for the images stored in the public folder while running on localhost. Same for the commented code in the test. But you're right, they aren't needed, I'll remove it.

File: app/models/article.rb
Comment (squarebat) [reply app/models/article.rb:32]: Yes I wanted to ask about how to tackle the hardcode url. Thanks for the info. If I am understanding it correctly, the current changes should work only for DEV, specifically for the images in articles, since the regex takes AWS values from the config. I think I could code the regex to take a storage_host and storage_dir from the config. Imgproxy images won't be detected however, their urls are structured differently, and depending on how they're stored, it might need separate deletion logic.

Comment (rhymes) [reply app/models/article.rb:32]: Cool! I like that we're moving in steps instead of trying to do it all at once! Thank you!

File: app/models/article.rb
Comment (maestromac) [inline app/models/article.rb:839]: Just thinking out loud: I wonder if it's better to limit each worker to only delete 1 image. Currently, it looks like we are throwing all images of an article to an array. If a worker failed and set to rerun, it would be trying to re-delete all the images again. 

I don't think there's a huge performance difference here but there's a benefit to being precise and strict with 1 arg instead of an array. Thoughts @rhymes ?

File: app/models/article.rb
Comment (squarebat) [reply app/models/article.rb]: I fixed the regex to take the path from store_dir. The regex will still detect only images stored in aws s3 though. Would it be a good idea to add a config variable like `AWS_S3_STORAGE_HOST = s3.amazonaws.com`, or perhaps create a few non aws specific config variables?

File: app/models/article.rb
Comment (rhymes) [reply app/models/article.rb:837]: Hi @maestromac, great point! Having idempotent workers is usually much better. If an article has to delete 5 images and the third fails, we'd be retrying in the first one. Given that the worker is super low priority, I'd rather have 5 workers each deleting one image.

Sorry to ask you this @squarebat but could you alter the code so that the worker only takes one path and instead submits N workers to the queue?

Comment (squarebat) [reply app/models/article.rb:837]: Yeah sure! That's actually a great idea.

File: app/models/article.rb
Comment (rhymes) [reply app/models/article.rb]: I'm not sure how we'll go about deleting images stored on imgproxy and as we said, this PR is about only DEV's Cloudinary images. I don't think we should add anything else

**Review by rhymes — APPROVED** [review]
Thanks for the changes @squarebat!

File: app/models/article.rb
Comment (jgaskins) [inline app/models/article.rb:30]: Do we prescribe S3 usage or do we allow for any S3-compatible storage, such as DigitalOcean Spaces or MinIO? If we _only_ allow S3, this looks like it should work, but if we allow any S3-compatible object storage, we may not want to hardcode S3 here.

/cc @rhymes

File: app/services/images/delete.rb
Comment (jgaskins) [inline app/services/images/delete.rb:21]: Do we need to retrieve the object before we remove it?

File: app/services/images/delete.rb
Comment (jgaskins) [inline app/services/images/delete.rb:22]: If this breaks halfway through (for example, a connection error to S3, on image 4 of 6), what happens when the job retries?

File: app/models/article.rb
Comment (msarit) [reply app/models/article.rb:30]: Seconded

File: app/services/images/delete.rb
Comment (maestromac) [reply app/services/images/delete.rb:21]: I can attest that this works since I [recommended it ](https://github.com/forem/forem/issues/13936#issuecomment-872514868)and tested it on Staging. I found suggestion [here](https://groups.google.com/g/carrierwave/c/15ALbubO5Uk).

File: app/models/article.rb
Comment (maestromac) [inline app/models/article.rb:838]: You are almost there. I'd you change how this method to accept a single string instead of an array. The same should be done with `Image::Delete` service

**Comment by squarebat** [comment]
`tagsFlows/tagIndexNavigation.spec.js` this e2e test keeps failing and then mysteriously passing for some unknown reason. Am I missing something that needs to be done?

File: app/models/article.rb
Comment (rhymes) [reply app/models/article.rb:30]: @jgaskins I think this PR is only intended to work on S3 and Cloudinary. I really don't know how other compatible storages can work as our entire configuration is based on S3:

https://github.com/forem/forem/blob/77885e8fa4c473f56ec20d273613ca751e22a78e/config/initializers/carrierwave.rb#L24

cc @maestromac

File: app/services/images/delete.rb
Comment (rhymes) [reply app/services/images/delete.rb]: @jgaskins the worker has been changed to accept only one image, so it's 100% idemponent. Each worker will work on a single image. See https://github.com/forem/forem/pull/14141/commits/6ac63c899733fb9507e8b60931c9f4a0e6042ae5

**Comment by rhymes** [comment]
@squarebat I think it's flakey, I will signal this internally and restart the build :) Thanks for keeping the branch updated!

File: app/models/article.rb
Comment (jgaskins) [reply app/models/article.rb:30]: _TL;DR: I'd like to get a lot more clarity on this from Product before we merge._

@rhymes DigitalOcean Spaces, MinIO, and at least a half-dozen other object-storage providers are API-compatible with S3. When I say API-compatible, I mean you could even use AWS's own S3 client to talk to them. The only thing you have to change is to provide the `endpoint` parameter:

If you need to connect to a non-standard endpoint, you may specify the :endpoint option

<cite>https://docs.aws.amazon.com/sdk-for-ruby/v3/api/</cite>

It's literally that simple. Fog supports [this `:endpoint` option](https://github.com/fog/fog-aws/blob/1210edde0fbff8f508ac60ebcfc0c7ca6136d678/lib/fog/aws/storage.rb#L435-L451), as well. I don't know if we provide that, but if not it'd be trivial to add. However, it does mean we need to think about whether we're forcing people to use S3 specifically or giving options for alternative, API-compatible object storage. This can be important for folks hosting a Forem instance inside their own cluster since MinIO can run in any containerized environment.

All this may sound like it's outside the scope of this PR, but I assure you it's not. 🙂 If we assume everyone's using S3 then we are implicitly deciding they must use S3. I don't know of any product guidance one way or the other, so maybe we should talk about it with Product before we make a decision that I don't think is ours to make. If someone is currently using S3-compatible storage that isn't AWS S3 and this gets merged in, these requests may fail for them and cause problems.

File: app/services/images/delete.rb
Comment (jgaskins) [reply app/services/images/delete.rb]: @maestromac I have no doubt it works. I did not mean to imply it wouldn't. 🙂 I mean do we _need_ to fetch the file before deleting it or can we simply delete it? For example, we can delete a file on disk if we know its path without reading it. If the file is in remote storage, we can issue a `DELETE` request without preceding it with a `GET` request.

Retrieving the file feels like unnecessary overhead here, so I'm wondering if CarrierWave requires that we do that or if we can skip that step.

File: app/models/article.rb
Comment (rhymes) [reply app/models/article.rb:30]: Thanks for the context! I do not know what's the answer here as this wasn't definitely on the radar when the issue was discussed and I'm not familiar about self host setups. Can you bring it to the attention of the Systems and Product team and follow up? This way @squarebat doesn't continue working on this PR before we have a final decision about what we want to do...

File: app/services/images/delete.rb
Comment (maestromac) [reply app/services/images/delete.rb]: After diving a bit into CarrierWave's API. I can't seem to find a way to do it without retrieving the file first :confused:

**Comment by maestromac** [comment]
Thank you again for this PR @squarebat . Unfortunately, we're going to have to close this PR.

We just realized that this feature would be introducing a security vulnerability; you can indirectly any user's article's image by referring to it on your article then delete it. There's also the product question of hosting service compatibility that needed to be worked out. 

Because you gave this a try, it made it easier for us to see the potential problems. Thank you!

**Comment by citizen428** [comment]
> @maestromac if we need to implement this in the future, we could do it with a reference count, where we count the number of times an image link is referred to and delete the image when count reaches zero.

A malicious party could create a draft post that holds a reference to every image (this could be automated) effectively preventing them all from ever getting removed. It's not exactly a major issue of course, since the result would be the same as if we did no deletions in the first place but if I learned anything from 20 years on working on the web is that whatever can be abused will be 😂

**Pull request closed without merging**
